### PR TITLE
Reapply "[Codegen][GPU] Add range information to GPU dispatch IDs" (#19361)

### PR DESCRIPTION
**Pull request by krzysz00**
This reverts commit cb5be1dbd3560f692578c137eadbb413b41e44c7.

Compaled to the previous revision, this one works around a correctness
bug in dataflow analysis that's being fixed by removing the analysis
after SCF->CF.

---

First, this patch implements InferIntRangeInterface for hal.interface.workgroup.{size,id,count} using a local upper_bound attribute.

Then, it adds a -iree-codegen-gpu-propagate-dispatch-size-bounds pass that adds these upper_bounds identifiers to the interface.workgroup operations and to gpu.thread_id based on static information available late in the codegen pipeline.

Then, it uses -optimize-int-arithmetic to optimize indexing after -lower-affine, getting rid of a bunch of "if the input's negative" logic that isn't actually needed in many of our kernels.

It also ensures that these upper_bound values propagate to LLVM.

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td [hunk 1]
```diff
@@ -3029,9 +3029,28 @@ def OpGroupInterfaceOps : OpDocGroup {
 
 let opDocGroup = OpGroupInterfaceOps in {
 
-def HAL_InterfaceWorkgroupIDOp : HAL_PureOp<"interface.workgroup.id", [
-  DeclareOpInterfaceMethods<OpAsmOpInterface, ["getAsmResultNames"]>,
-]> {
+class HAL_InterfaceWorkgroupOp<string mnemonic, list<Trait> traits = []>
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/HAL/IR/HALOps.td:3032]: nice cleanup!

**Review by benvanik — APPROVED** [review]
LGTM for HAL changes! may want @MaheshRavishankar or someone to take a peek at the codegen side.

**Review by qedawkins — APPROVED** [review]
IIRC this got reverted for a regression. If the regression is fixed LGTM!

**Review by kuhar — COMMENTED** [review]
Could explain what fixed the regression that caused the initial revert? It would be good to include this in the PR description.

**Pull request merged**